### DISCUSSION TITLE
posts/react-navigation-without-react-router/

### DISCUSSION THREAD

**Discussion by giscus** [ncoughlin/ncoughlin-giscus#4] [Announcements]
# posts/react-navigation-without-react-router/

The theory behind navigation in a React application without using React-Router

https://ncoughlin.com/posts/react-navigation-without-react-router/?utm_source=chatgpt.com

**Comment by Danea1-oss** [comment]
import { useState as usePathState, useEffect as usePathEffect } from "react";
import LandingPage from "./pages/HomePage.jsx";
import AuthPage from "./pages/LoginPage.jsx";
import FilmsPage from "./pages/MoviesPage.jsx";
import NotFoundPage from "./pages/ErrorPage.js";
import FilmDetailsPage from "./pages/MovieDetailsPage.js";
import HeaderNavigation from "./components/MainNavigation.js";

function MainApp() {
  const [currentRoute, setCurrentRoute] = usePathState(window.location.pathname);

  const handleNavigation = (route) => {
    window.history.pushState({}, "", route);
    setCurrentRoute(route);
  };

  usePathEffect(() => {
    const updateRouteOnPopState = () => {
      setCurrentRoute(window.location.pathname);
    };

    window.addEventListener("popstate", updateRouteOnPopState);
    return () => window.removeEventListener("popstate", updateRouteOnPopState);
  }, []);

  let ActiveComponent;
  // Check the current route
  if (currentRoute === "/") {
    ActiveComponent = LandingPage;
  } else if (currentRoute === "/login") {
    ActiveComponent = AuthPage;
  } else if (currentRoute === "/movies") {
    ActiveComponent = () => <FilmsPage navigate={handleNavigation} />;
  } else if (currentRoute.startsWith("/movies/")) {
    ActiveComponent = () => <FilmDetailsPage />;
  } else {
    ActiveComponent = NotFoundPage;
  }

  return (
    <div>
      <HeaderNavigation navigate={handleNavigation} />
      <ActiveComponent navigate={handleNavigation} />
    </div>
  );
}

export default MainApp;
